endpoint / BPF: add dynamic configuration for endpoint for default-all lookup upon ingress

ianvernon opened the pull request:
The below is pasted from #6047 : 

```release-note
Fix issue where Identity propagation delay may cause endpoints with no policy to drop packets from newly created endpoints.
```
The goal of this is to allow configuration which would otherwise be implemented via preprocessor macros to instead be configurable at runtime. Over time, this will allow recompilation of endpoints to be largely eliminated.

To allow the endpoint to reach into a configuration map value that is specific to the endpoint, it could be implemented in one of a couple of ways:
1. One big map of config options indexed by `LXC_ID` and sized based on the number of anticipated endpoints.
1. One small single-element array per endpoint.

Option (1) above would require the `LXC_ID` to be specified at compile time, with a single static map path for configuration of all endpoints. This `LXC_ID` would be embedded in the BPF instructions for the program as a register initialization instruction prior to a `BPF_CALL` instruction that calls the map lookup API. If we wanted to templatize this field then we would need to build ELF parsing functionality that is able to find each map call function in the instructions and correlate it with the corresponding map, then update the register initialization prior to the lookup for the options map.

**UPDATE**: With some simple macros, a 3-line iproute2 change and global variables, we could easily achieve the above (see http://vger.kernel.org/lpc-bpf2018.html#session-3).

Option (2) would require the map's path to be specified at compile time. This is embedded into 3 ELF sections which are easier to identify for templatization: Any program section that refers to the map; the debuginfo for the map; and the symbol table.

Given that Option 2 has more potential for a path to reduce BPF program compilation to O(1) rather than O(N) for the number of endpoints N on the node, it seems like a better option.

For now, particularly given that there is potential for this to be backported, this PR only attempts to implement `POLICY_INGRESS` skipping when policy is disabled for the direction.

Follow-up work is tracked by #6273

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6208)
<!-- Reviewable:end -->

joestringer committed:
maps: Add BPFConfigMap for endpoint configuration

Signed-off-by: Joe Stringer <joe@cilium.io>
Signed-off by: Ian Vernon <ian@cilium.io>

joestringer committed:
endpoint: Support dynamic BPF configuration

Signed-off-by: Joe Stringer <joe@cilium.io>
Signed-off by: Ian Vernon <ian@cilium.io>

joestringer committed:
bpf: Templatize endpoint configuration

Signed-off-by: Joe Stringer <joe@cilium.io>
Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
cmd: add `cilium bpf config get` command

This allows for direct introspection into endpoint configuration maps in BPF.

Signed-off by: Ian Vernon <ian@cilium.io>

ianvernon commented on bpf/bpf_lxc.c:
```diff
@@ -209,17 +210,22 @@ static inline int ipv6_l3_from_lxc(struct __sk_buff *skb,
+	if (!(cfg->flags & EP_F_SKIP_POLICY_EGRESS)) {
+		/* If the packet is in the establishing direction and it's destined
+		 * within the cluster, it must match policy or be dropped. If it's
+		 * * bound for the host/outside, perform the CIDR policy check. */
```
Remove extra ampersand.

ianvernon commented on bpf/bpf_lxc.c:
```diff
@@ -209,17 +210,22 @@ static inline int ipv6_l3_from_lxc(struct __sk_buff *skb,
 	if (ret != CT_REPLY && ret != CT_RELATED && verdict < 0) {
 		/* If the connection was previously known and packet is now
-		 * denied, remove the connection tracking entry */
+		 * * denied, remove the connection tracking entry */
```
Remove extra ampersand.

ianvernon commented on bpf/bpf_lxc.c:
```diff
@@ -510,16 +522,18 @@ static inline int handle_ipv4_from_lxc(struct __sk_buff *skb, __u32 *dstID)
 			   orig_dip, *dstID);
 	}
 
+	// TODO (ianvernon) - plumb cfg check here. can't until code is simplified
```
This still occurs, and results in the verifier complaining:

```
Prog section '2/7' rejected: Invalid argument (22)!
 - Type:         3
 - Instructions: 4118 (22 over limit)
 - License:      GPL

Verifier analysis:

Error filling program arrays!
```

ianvernon commented on bpf/bpf_lxc.c:
```diff
@@ -672,9 +686,13 @@ static inline int handle_ipv4_from_lxc(struct __sk_buff *skb, __u32 *dstID)
 
 __section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_IPV4_FROM_LXC) int tail_handle_ipv4(struct __sk_buff *skb)
 {
+    struct ep_config *cfg = lookup_ep_config();
```
Sigh, I think my spacing is off. @joestringer what is the appropriate spacing for the C code?

joestringer replied:
@ianvernon It's tabs indented 8 characters, just like our Go code.

ianvernon replied:
I think I caught all of the irregular whitespace.

ianvernon commented on bpf/lib/maps.h:
```diff
@@ -72,6 +72,13 @@ struct bpf_elf_map __section_maps POLICY_MAP = {
+	.type		= BPF_MAP_TYPE_ARRAY,
+	.size_key	= sizeof(__u32),
+	.size_value	= sizeof(struct ep_config),
+	.max_elem	= 2,
```
Change back to 1.

ianvernon commented on cilium/cmd/bpf_config_get.go:
```diff
@@ -0,0 +1,83 @@
+	lbl := args[0]
+
+	if lbl == "" {
+		Fatalf("Need ID or label\n")
```
Change message to only refer to endpoint ID.

ianvernon commented on cilium/cmd/bpf_config_get.go:
```diff
@@ -0,0 +1,83 @@
+	if lbl == "" {
+		Fatalf("Need ID or label\n")
+	}
+
```
Add a feature flag for this command, since it won't work until 
https://github.com/cilium/cilium/issues/6228 is fixed.

joestringer replied:
What's the intent of the feature flag exactly?

ianvernon commented on pkg/endpoint/endpoint.go:
```diff
@@ -1250,6 +1260,15 @@ func mapPath(mapname string, id int) string {
+	return mapPath(bpfconfig.MapNamePrefix, int(e.ID))
+}
+
+func (e *Endpoint) BPFConfigMapPathWithoutPrefix() string {
```
I will add this. The reason I need this is for when we open the map for dumping in the CLI command `cilium bpf config get <endpoint ID>`.

ianvernon commented on pkg/maps/configmap/doc.go:
```diff
@@ -0,0 +1,19 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package configmap represents the endpoints configuration. It is
```
endpoint's

ianvernon commented on pkg/maps/configmap/configmap.go:
```diff
@@ -0,0 +1,213 @@
+	MapNamePrefix = "cilium_ep_config_"
+
+	// MaxEntries represents the maximum number of elements in the map
+	MaxEntries = 2
```
Change back to 1.

ianvernon commented on test/k8sT/istio.go:
```diff
@@ -17,7 +17,6 @@ package k8sTest
 	"context"
 	"fmt"
-
 	. "github.com/cilium/cilium/test/ginkgo-ext"
```
Re-add this.

ianvernon replied:
Removed Istio test-related commits to reduce size of this PR.

ianvernon commented on test/k8sT/istio.go:
```diff
@@ -46,6 +45,11 @@ var _ = Describe("K8sIstioTest", func() {
 		// Istio 1.0.0. mTLS is enabled.
 		istioYAMLPath = helpers.ManifestGet("istio-cilium.yaml")
 
+		// istioProxyImage is the name of the Cilium Istio proxy image.
```
Need to run this against the Istio tests.

ianvernon replied:
Removed Istio test-related commits to reduce size of this PR.

aanm commented on pkg/maps/configmap/configmap.go:
```diff
@@ -27,5 +27,187 @@ import (
+// otherwise an error.
+func OpenMap(path string) (*EndpointConfigMap, bool, error) {
+
+	newMap := bpf.NewMap("cilium_lb4_services",
```
Is this supposed to be the right name? :thinking:

ianvernon replied:
This function is dead code, sorry about that. I'll remove it in the next upload.

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -613,6 +624,18 @@ func (e *Endpoint) regenerateBPF(owner Owner, currentDir, nextDir string, regenC
+		// inconsistency issues as above if there is a failure. Long
+		// term the solution to this is to templatize this map in the
+		// ELF file, but there's no solution to this just yet.
+		if err = e.bpfConfigMap.Update(e.desiredBPFConfig); err != nil {
```
This is weird because we are doing this update and everyone is happy. What happens if line 642 fails? Are we reverting this update?

ianvernon replied:
I pushed a function onto the `RevertStack` in the latest upload.

aanm commented on pkg/endpoint/endpoint.go:
```diff
@@ -290,6 +291,15 @@ type Endpoint struct {
 	desiredMapState PolicyMapState
 
+	// BPFConfigMap provides access to the endpoint's BPF configuration.
+	bpfConfigMap *bpfconfig.EndpointConfigMap
```
Maybe we should start thinking on having a `desiredState` structure and a `realizedState` structure that contains all of these maps inside of that structure.

ianvernon replied:
Agreed.

ianvernon replied:
FYI I am working on something like this that is coming soon in an upcoming PR.

aanm commented on pkg/maps/configmap/configmap.go:
```diff
@@ -139,7 +139,7 @@ func GetConfig(e endpoint) *EndpointConfig {
 }
 
-// Sync pushes the configuration options from the specified endpoint into the
+// Update pushes the configuration options from the specified endpoint into the
```
Out of scope, it's easy to review PRs if they have less lines to read.

aanm commented on test/k8sT/istio.go:
```diff
@@ -101,6 +101,14 @@ var _ = Describe("K8sIstioTest", func() {
 		ProvisionInfraPods(kubectl)
 
+		By("Disabling debug mode for Istio test")
+		err := kubectl.CiliumExecAll("cilium config Debug=Disabled")
```
Why? It will be hard to debug in case something fails.

ianvernon replied:
Debug mode puts more load on the system. While I was running the Istio tests (which revealed the bug that this PR is fixing), I was encountering severely high load (values up to 42). I can drop these Istio CI commits and work exclusively on fixing the Istio tests in another branch if you like.

ianvernon replied:
Saved the commits in a separate branch and dropped them from this PR.

Commit:
cmd: disable `cilium bpf config get` for now

Due to GH-6228, disable the direct dumping of BPF endpoint configuration maps.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
endpoint: revert endpoint BPF config map update if regenerateBPF fails

Signed-off by: Ian Vernon <ian@cilium.io>

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -209,17 +210,22 @@ static inline int ipv6_l3_from_lxc(struct __sk_buff *skb,
+		verdict = policy_can_egress6(skb, tuple, *dstID,
+				ipv6_ct_tuple_get_daddr(tuple));
+	}
+	else {
```
nit: With braces, the line should be `} else {`. With single-line contents inside each condition, the datapath code usually omits the braces entirely.

ianvernon replied:
I'm moving the check for IPv6 egress for consistency with IPv4 egress not having this. Will be fixed in next upload.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -510,16 +522,18 @@ static inline int handle_ipv4_from_lxc(struct __sk_buff *skb, __u32 *dstID)
 	 * bound for the host/outside, perform the CIDR policy check. */
-	verdict = policy_can_egress4(skb, &tuple, *dstID, ipv4_ct_tuple_get_daddr(&tuple));
+	verdict = policy_can_egress4(skb, &tuple, *dstID, ipv4_ct_tuple_get_daddr(&tuple)) ;
+
```
nit: This git hunk doesn't do anything, should be dropped.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -510,16 +522,18 @@ static inline int handle_ipv4_from_lxc(struct __sk_buff *skb, __u32 *dstID)
 		/* If the connection was previously known and packet is now
 		 * denied, remove the connection tracking entry */
 		if (ret == CT_ESTABLISHED)
 			ct_delete4(get_ct_map4(&tuple), &tuple, skb);
```
nit: Unnecessary whitespace change here

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -672,9 +686,13 @@ static inline int handle_ipv4_from_lxc(struct __sk_buff *skb, __u32 *dstID)
 	__u32 dstID = 0;
-	int ret = handle_ipv4_from_lxc(skb, &dstID);
-
+	int ret ;
```
nit: No whitespace necessary between the variable name and semicolon.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -800,19 +818,23 @@ ipv6_policy(struct __sk_buff *skb, int ifindex, __u32 src_label, int *forwarding
+				tuple.nexthdr, sizeof(tuple.saddr),
+				&tuple.saddr, false);
+	}
+	else {
```
nit again, else check doesn't follow usual conventions in these files.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -800,19 +818,23 @@ ipv6_policy(struct __sk_buff *skb, int ifindex, __u32 src_label, int *forwarding
+         	 * denied, remove the connection tracking entry */
+        	if (ret == CT_ESTABLISHED)
+            		ct_delete6(get_ct_map6(&tuple), &tuple, skb);
+			return verdict;
```
nit: Whitespace is spaces here. (That's also why the `permitted by policy` line above reports a change)

joestringer replied:
Additional nit, this messes up the indentation of the verdict return. The if statement deliberately only applies to the `ct_delete6()` line (no braces => only apply to the next statement), however all traffic that is rejected by the verdict and not reply / not related is dropped by this verdict return.

Is there any intentional change in these lines? Seems like the hunk could just be dropped.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -866,7 +888,13 @@ __section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_IPV6_TO_LXC) int tail_ipv6_policy(s
-	ret = ipv6_policy(skb, ifindex, src_label, &forwarding_reason);
+	struct ep_config *cfg = lookup_ep_config();
+
+	if(cfg)
```
nit: space between `if` and open brace.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -866,7 +888,13 @@ __section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_IPV6_TO_LXC) int tail_ipv6_policy(s
+	struct ep_config *cfg = lookup_ep_config();
+
+	if(cfg)
+	    ret = ipv6_policy(skb, ifindex, src_label, &forwarding_reason, cfg);
```
nit: spacing

(You can likely search for instances of 2 or more spaces and they likely should be substituted for tabs)

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -934,9 +962,13 @@ ipv4_policy(struct __sk_buff *skb, int ifindex, __u32 src_label, int *forwarding
+						    sizeof(orig_sip),
+						    &orig_sip, is_fragment);
+	//else
+	//	verdict = 0;
```
Is this deliberately omitted? Is it because we can trust the verdict to be 0 if we don't run the policy access function? In that case we can likely remove the commented else lines here, and also make equivalent changes in the corresponding ipv6 case so that the code is logically the same for both cases.

ianvernon replied:
This was leftover from me trying to reduce complexity (and failing). I'll uncomment it.

joestringer commented on pkg/maps/configmap/configmap.go:
```diff
@@ -0,0 +1,179 @@
+			if flag, ok := flagsToString[bitVal]; ok {
+				buffer.WriteString(flag)
+				buffer.WriteString(",")
+			}
```
Not a biggie, but `} else { buffer.WriteString("%04x", 1 << i) }` ?

Ie, if it can't be found in the map to convert it to a string, then print the raw bit value.

This would mean that if we somehow upgrade/downgrade and don't change the map (so it could have bit values not understood by this userspace), then it would at least print that the underlying map contains a non-zero bitvalue for this offset.

joestringer reviewed (changes requested):
I have a whole bunch of mostly style nits where the C code does not follow the same conventions in the rest of the file. Some of these can be hightlighted by running `git show`, others are less obvious. This includes:
* Git hunks that appear to do nothing (no need to change lines that aren't logically modified)
* Still a few spaces where there should be tabs
* if/else conditions usually have no braces if there is exactly one nested statement, or braces if there are more. The brace for else should be `} else {` all on one line.
* Spaces before the end-of-line semicolon.
* Adding commented-out code

More importantly, for your question that's relevant to the sizing of the map type, by my counting the endpoint configuration currently tracks:
* 6B of NODE_MAC
* 16B of IPv6 address
* 4B of IPv4 address
* 2B of LXC_ID
* 2B of LXC_ID in network byte-order
* 4B of SECLABEL
* 4B of SECLABEL in network byte-order
* Eleven configuration options. Given one bit for each, these fit into 2B. Reserving 4 contiguous bytes for it should be good, like this PR does today.

This gives us about 40B total configuration. We are generating a map per endpoint so given that we won't overprovision for the maximum number of endpoints we may ever need, it's fairly low overhead. 40B safely covers what we need today, we could perhaps pad that out to something like 48B to give us a little more space to deal with just in case.

We currently expect to move over time to retire this map, but in case we don't it'd be nice to atleast handle some amount of extensibility to use all of the fields we have today plus some extra space. I could see a reasonable argument that 8B extra is not enough, but I think that even in that case we still have some flexibility in tweaking what we hold in the config map, particularly given that we are only relying on 2 bits in this PR. Worst case if we really want to change the size in future, we can always do this disruptively by just changing the size, or if it's really worthwhile, come up with some nifty map migration mechanism.

ianvernon commented on bpf/lib/common.h:
```diff
@@ -509,6 +510,19 @@ struct proxy6_tbl_value {
 } __attribute__((packed));
 
+/* ep_config corresponds to the EndpointConfig object in pkg/maps/configmap. */
+struct ep_config {
```
@joestringer PTAL

joestringer replied:
Apparently I missed that the MAC types are actually 8B in the tree, so all-up the non-pad fields below add to 44B. I guess we can round to the 8B border, so pad should be 4B rather than 1 as below.

ianvernon replied:
Fixed to be uint32 in both Cilium code and BPF.

ianvernon commented on pkg/maps/configmap/configmap.go:
```diff
@@ -19,13 +19,176 @@ import (
+func (k *Key) NewValue() bpf.MapValue { return &EndpointConfig{} }
+
+func (cfg *EndpointConfig) String() string {
+	return fmt.Sprintf("%s,%d,%d,%s,%s,%d,%d,%s", cfg.Flags.String(), cfg.SecurityIdentity, cfg.SecurityIdentityNB, cfg.IPv4.String(), cfg.IPv6.String(), cfg.LXCID, cfg.LXCIDNB, cfg.NodeMAC.String())
```
@joestringer I'm ambivalent about whether we want to write all the data here since we aren't plumbing it yet. Open to suggestions.

joestringer replied:
The string output format is not considered stable and is liable to change anyway, so I'd probably just put what we have so far. 

From a CLI consistency point of view, I'd probably want this to be space (tab?) delimited, I think that with tabs and the TabWriter we can end up with nicely formatted columns. Again, just a nit but if you end up changing this anyway then it may be worth considering.

ianvernon replied:
I put this as a TODO comment, but added spaces in the meantime.

joestringer commented on pkg/maps/configmap/configmap.go:
```diff
@@ -0,0 +1,195 @@
+// EndpointConfig represents the value of the endpoint's BPF map.
+//
+// Must be in sync with struct config_value in <bpf/lib/common.h>
+type EndpointConfig struct {
```
Do we need to add some code somewhere for verifying that this structure matches the C version?

ianvernon replied:
Did this in `pkg/alignchecker` in newest commit.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -800,9 +800,13 @@ ipv6_policy(struct __sk_buff *skb, int ifindex, __u32 src_label, int *forwarding
+				tuple.nexthdr, sizeof(tuple.saddr),
+				&tuple.saddr, false);
+	} else {
+		verdict = 0 ;
```
nits: unnecessary braces, space before colon.

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -866,7 +869,13 @@ __section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_IPV6_TO_LXC) int tail_ipv6_policy(s
+	if (cfg)
+		ret = ipv6_policy(skb, ifindex, src_label, &forwarding_reason, cfg);
+	else
+		ret = DROP_NO_CONFIG ;
```
nit: space before colon

joestringer reviewed (approved):
Changes LGTM, there's still a couple of minor nits below that I missed earlier.

Commit:
alignchecker: add EndpointConfig validation

Signed-off by: Ian Vernon <ian@cilium.io>

aanm commented on pkg/maps/configmap/configmap.go:
```diff
@@ -15,12 +15,181 @@
+}
+
+func (k Key) String() string {
+	return fmt.Sprintf("%d", k)
```
`return fmt.Sprintf("%d", k.Bits)`

ianvernon replied:
Will be fixed in next upload.

aanm commented on pkg/maps/configmap/configmap.go:
```diff
@@ -15,12 +15,181 @@
+// Must be in sync with struct config_value in <bpf/lib/common.h>
+type EndpointConfig struct {
+	Flags Flags
+	// NOTE: the below are not plumbed yet. They are for future work.
```
Which below? All of configuration or `IPv4/IPv6`?

joestringer replied:
All of it. If you'd prefer, it could be swapped for a bunch of placeholder bytes but I don't really mind either way.

ianvernon replied:
I'll have the comment updated and a GitHub issue reference in the next upload.

aanm commented on pkg/maps/configmap/configmap.go:
```diff
@@ -15,12 +15,181 @@
+}
+
+// GetConfig creates a EndpointConfig structure using the endpoint's
+// configuration.
```
Add comment the endpoint needs to have its mutex held.

ianvernon replied:
Comment will be present in next upload.

aanm commented on pkg/maps/configmap/configmap.go:
```diff
@@ -15,12 +15,181 @@
+			if err := bpf.ConvertKeyValue(key, value, &k, &v); err != nil {
+				return nil, nil, err
+			}
+			fmt.Printf("k: %d, v: %b", k, v.Flags)
```
Is this `fmt.Printf` intended to be kept?

ianvernon replied:
Will be removed in next upload.

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -103,6 +103,16 @@ func (e *Endpoint) CallsMapPathLocked() string {
 }
 
+// BPFConfigMapPathLocked returns the path to the BPF config map of endpoint.
+func (e *Endpoint) BPFConfigMapPathLocked() string {
```
Why does it need to be "`Locked`"? The endpoint ID doesn't require mutex to be held.

ianvernon replied:
Will be called BPFConfigMapPath in next upload.

aanm commented on bpf/bpf_lxc.c:
```diff
@@ -800,9 +800,12 @@ ipv6_policy(struct __sk_buff *skb, int ifindex, __u32 src_label, int *forwarding
+				tuple.nexthdr, sizeof(tuple.saddr),
+				&tuple.saddr, false);
+	else
+		verdict = 0;
```
`verdict = TC_ACT_OK;`

aanm commented on bpf/bpf_lxc.c:
```diff
@@ -811,7 +814,6 @@ ipv6_policy(struct __sk_buff *skb, int ifindex, __u32 src_label, int *forwarding
 		 * denied, remove the connection tracking entry */
 		if (ret == CT_ESTABLISHED)
 			ct_delete6(get_ct_map6(&tuple), &tuple, skb);
-
```
Accidental delete?

aanm commented on bpf/bpf_lxc.c:
```diff
@@ -866,7 +868,13 @@ __section_tail(CILIUM_MAP_CALLS, CILIUM_CALL_IPV6_TO_LXC) int tail_ipv6_policy(s
+	if (cfg)
+		ret = ipv6_policy(skb, ifindex, src_label, &forwarding_reason, cfg);
+	else
+		ret = DROP_NO_CONFIG;
```
Will the monitor notify us the packets were dropped due no configuration? cc @joestringer

joestringer replied:
Yes, that's the lines immediately below which check whether `ret` is an error and perform `send_drop_notify()`.

aanm commented on bpf/bpf_lxc.c:
```diff
@@ -934,9 +942,13 @@ ipv4_policy(struct __sk_buff *skb, int ifindex, __u32 src_label, int *forwarding
+						    sizeof(orig_sip),
+						    &orig_sip, is_fragment);
+	else
+		verdict = 0;
```
verdict = TC_ACT_OK;

aanm commented on bpf/bpf_lxc.c:
```diff
@@ -995,11 +1007,15 @@ ipv4_policy(struct __sk_buff *skb, int ifindex, __u32 src_label, int *forwarding
+	if (cfg)
+		ret = ipv4_policy(skb, ifindex, src_label, &forwarding_reason, cfg);
+	else
+		ret = DROP_NO_CONFIG;
```
Will the monitor notify us the packets were dropped due no configuration? cc @joestringer

joestringer replied:
yes, see https://github.com/cilium/cilium/pull/6208/files#r235173998.

aanm commented on cilium/cmd/bpf_config_get.go:
```diff
@@ -33,6 +33,11 @@ var bpfConfigGetCmd = &cobra.Command{
 		common.RequireRootPrivilege("cilium bpf config get")
 		requireEndpointID(cmd, args)
+
+		if true {
```
Leftover?

ianvernon replied:
This is intentional. The `PrintF` statement elaborates on why this is done.

raybejjani reviewed (approved):
approved for docs and monitor

aanm commented on bpf/bpf_lxc.c:
```diff
@@ -800,9 +800,12 @@ ipv6_policy(struct __sk_buff *skb, int ifindex, __u32 src_label, int *forwarding
-	verdict = policy_can_access_ingress(skb, src_label, tuple.dport,
-					    tuple.nexthdr, sizeof(tuple.saddr),
-					    &tuple.saddr, false);
+	if (!(cfg->flags & EP_F_SKIP_POLICY_INGRESS))
```
Can be added has an optimization later, @joestringer would the `unlikely` help here? (As it is more likely an endpoint to have policy enforced)

joestringer replied:
Seems reasonable, but agree that it doesn't need to block the PR.

aanm commented on bpf/bpf_lxc.c:
```diff
@@ -934,9 +943,13 @@ ipv4_policy(struct __sk_buff *skb, int ifindex, __u32 src_label, int *forwarding
-	verdict = policy_can_access_ingress(skb, src_label, tuple.dport,
-					    tuple.nexthdr, sizeof(orig_sip),
-					    &orig_sip, is_fragment);
+	if (!(cfg->flags & EP_F_SKIP_POLICY_INGRESS))
```
Same here for the `unlikely`
